src/transformers/models/dbrx/modeling_dbrx.py

```diff
@@ -0,0 +1,2132 @@
+#############################################################################
+
+
+def _get_unpad_data(attention_mask: torch.Tensor):
+    seqlens_in_batch = attention_mask.sum(dim=-1, dtype=torch.int32)
+    indices = torch.nonzero(attention_mask.flatten(), as_tuple=False).flatten()
+    max_seqlen_in_batch = seqlens_in_batch.max().item()
+    cu_seqlens = F.pad(torch.cumsum(seqlens_in_batch, dim=0, dtype=torch.int32),
+                       (1, 0))
+    return (
+        indices,
+        cu_seqlens,
+        max_seqlen_in_batch,
+    )
+
+
+class DbrxAttention(nn.Module):
+    """Multi-head self attention."""
+
+    def __init__(self,
+                 hidden_size: int,
+                 num_heads: int,
+                 max_position_embeddings: int,
+                 attn_config: DbrxAttentionConfig,
+                 block_idx: Optional[int] = None):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.num_heads = num_heads
+        self.head_dim = self.hidden_size // self.num_heads
+        self.max_position_embeddings = max_position_embeddings
+        self.block_idx = block_idx
+        self.config = attn_config
+        if block_idx is None:
+            logger.warning_once(
+                f'Instantiating {self.__class__.__name__} without passing a `block_idx` is not recommended and will '
+                +
+                'lead to errors during the forward call if caching is used. Please make sure to provide a `block_idx` '
+                + 'when creating this class.')
+
+        self.attn_pdrop = attn_config.attn_pdrop
+        self.clip_qkv = attn_config.clip_qkv
+        self.num_key_value_heads = attn_config.kv_n_heads
+        self.num_key_value_groups = self.num_heads // self.num_key_value_heads
+        self.rope_theta = attn_config.rope_theta
+
+        self.Wqkv = nn.Linear(self.hidden_size,
+                              self.hidden_size +
+                              2 * self.num_key_value_heads * self.head_dim,
+                              bias=False)
+        self.out_proj = nn.Linear(self.hidden_size,
+                                  self.hidden_size,
+                                  bias=False)
+        self.rotary_emb = DbrxRotaryEmbedding(
+            self.head_dim,
+            max_position_embeddings=self.max_position_embeddings,
+            base=self.rope_theta,
+        )
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        position_ids: torch.LongTensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        past_key_value: Optional[Cache] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        **kwargs: Any,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Cache]]:
+        bsz, q_len, _ = hidden_states.size()
+
+        qkv_states = self.Wqkv(hidden_states)
+        if self.clip_qkv is not None:
```

Comment:
Is `clip_qkv` always used in the released model? if yes we can remove code path

Reply:
I think we'd like to keep it configurable, sometimes we ablate the clipping value and though it's possible to just set it to a very large value (to nullify behavior), it seems better to be able to set it to `None`.

Reply:
Understood, but the model that you release does not change it and having these kind of arguments in the config that change the forward is not much to our liking! That's why we'd rather have it removed

Reply:
Ah ok, I'm guessing this interferes with `torch.compile`? Let's see...

would this be ok? 
```python
min = -self.clip_qkv if self.clip_qkv is not None else None
max = self.clip_qkv
qkv_states = qkv_states.clamp(min=min, max=max)
```

This avoids the conditional branch on the clamp, and it seems that `torch.clamp` can accept `None` values: https://pytorch.org/docs/stable/generated/torch.clamp.html

again the reasoning here is that Dbrx is not just one model, we expect folks to train and finetune lots of copies of this model arch, and we want some way of turning clipping OFF, so either we need to put it in a very large value like 1e10 or allow `None`